### ISSUE TITLE
Make CanJS tree-shakable

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/canjs#4013]
__TLDR:__ Make it so the `can` package export is tree-shake-able. Allowing you to import just what you want like:

```js
import {Component, observe} from "can"
```

> This was discussed on a [recent live stream (8:14)](https://youtu.be/EEBytcIX7oA?t=8m14s).

## The Problem 

Currently, to use CanJS, you likely have to install at least 8 different packages:

```
npm install can-component can-connect can-define can-route can-route-pushstate can-set can-stache can-stache-bindings --save
```

While this is for good reason, it's annoying.

Furthermore, to build something like a model, you need to import a bunch of stuff:

```js
var DefineMap = require("can-define/map/map");
var DefineList = require("can-define/list/list");
var superMap = require("can-connect/can/super-map/super-map")
var set = require("can-set");
```

## The Solution

I'd like to make the `can` package's main export a tree-shake-able `ES` module.  This would allow importing enough to make a model look like:

```js
import {DefineMap, DefineList, superMap, set} from "can";
```

This module should be tree-shake-able by both webpack and stealjs.  So if someone imports only, for example `ajax` like:

```js
import {ajax} from "can";
```

They would only get `ajax` in their build.

### What is the minimal viable example of this?

Ideally, we will not have to rewrite every existing project to `ES` modules.  Instead, `steal.export()` should be able to do the necessary transformations.

Also, it would probably be nice if we don't have to move every module within the single export.  The main module might look like:

```js
// can.js 
exports Component from "can-component";
exports route from "can-route";
```

Essentially, we will export everything by it's `can-namespace` name.  Perhaps there's a way to do this programmatically within the module, but doubt it.  Something like:

```js
import "moduleA";
import "moduleB";
...
import "moduleZ";
import can from "can-namespace";
for(var prop in can){
  exports [prop] as can[prop]
}
```

But I would be fine maintaining this list of modules and exports for now.

### How do we document this?

There are a few places in the docs that might be affected by using tree-shaking as the primary way of using CanJS:

- The sidebar
- The title
- Examples

I still think CanJS's sidebar should use the `module-name` listing.  But we should show the alias, perhaps next to the module name in the title area like:

<img width="879" alt="canjs_-_can-component" src="https://user-images.githubusercontent.com/78602/37106914-97a146fe-21f8-11e8-9f88-bd00be23a85e.png">

What should examples how?

Guides should almost certainly show using the `tree-shaking` module. This will make JSBin / script tag use very similar.  But what about individual packages and modules?

**Comment by matthewp** [comment]
Import options:

```js
import {DefineMap, DefineList, superMap, set} from "can";
```

vs.

```js
import can from "can";

can.Component ...
```

**Comment by matthewp** [comment]
Tested this out and if you have a module like:

```js
import { Component } from 'can/es';

console.log(Component);
```

Where __can/es__ looks like:

```js
export {
	default as connect
} from 'can-connect/all';

export {
	default as DefineMap
} from 'can-define/map/map';

export {
	default as Component
} from 'can-component/es';
```

It will tree-shake out can-connect. Notice that the can-component module is __can-component/es__. It looks like this:

```js
import Component from './can-component';

export default Component;
```

We need to add these wrapper modules to all of the projects that we want to include in `can/es`. We also need to add `"sideEffects": true` to the package.json of these projects.

**Comment by justinbmeyer** [comment]
@matthewp A few questions:

1.  Does it tree-shake out other packages?  So `can-set` isn't included?  What about modules in `can-util` that `can-connect` might use, but the others don't.
1.  Does this imply every module (not just package) needs an `es` counterpart?
2.  These are the requirements for Webpack I assume.  Do you know:
   1.  Are these technical/mathmatical requirements, or is it just Webpack needs things a certain way right now.  Is webpack looking to make CJS modules work?
   2.  What does `sideEffects` true do? What's it's purpose?

**Comment by matthewp** [comment]
1. It doesn't include can-set. Not sure about can-util types of things, as I don't know what can-connect uses that others do not.
2. No, only the modules that are imported by `can/es` would need the wrapper `es` module.
3.
  i. I don't know of a technical requirement at this time, I don't plan on making the wrapper a requirement but perhaps there is a good reason, will have to wait until I add it to steal-tools to know.
  ii. It's an opt-in that a package can be tree-shaken, since `import { foo } from 'pkg'` could cause code to break if tree-shaken if importing causes needed side-effects.

**Comment by justinbmeyer** [comment]
To clarify:

> No, only the modules that are imported by can/es would need the wrapper es module.

You mean only modules __DIRECTLY__ imported by `can/es` would need the wrapper ... right?

**Comment by matthewp** [comment]
I don't understand your question, `can-connect` is also not an ES module. Why are you surprised by `can-util/foo` but not `can-connect`?

**Comment by matthewp** [comment]
I just tested and the wrapper modules do not need be in the separate packages. So if we wanted we could add (within canjs) a `es/component.js` that acts as the wrapper:

```js
import Component from 'can-component';

export default Component;
```

**Comment by justinbmeyer** [comment]
> I don't understand your question, can-component is also not an ES module. Why are you surprised by can-util/foo but not can-component?

Yeah, it's suprising to me that it's able to tree shake those CJS modules too.

**Comment by matthewp** [comment]
It doesn't actually tree-shake the CJS module, it tree shakes the ES wrapper.  Think about it like this, you have code:

__app.js__

```js
import { Component } from 'can/es';
```

__can/es.js__

```js
export { default as Component } from 'can-component/es';
export { default as connect } from 'can-connect/es';
```

When the tree-shaking occurs the later gets treated as if it were:

```js
export { default as Component } from 'can-component/es';
```

At this point it doesn't need to understand the CJS modules of `can-connect`, it's like it was never even imported.

**Comment by matthewp** [comment]
This is actually really good news because it means we don't have to rewrite everything into ES modules. The question becomes should we:

1. Create wrapper modules in each package, ie create a `can-component/es` that just reexports the CJS source.
2. Create the wrapper modules in canjs/canjs. This prevents needing to spread this feature into each package.

Argument for (1) would be that it makes it easier for people that use the individual packages to still get benefits of tree-shaking.

**Comment by chasenlehara** [comment]
+1 on option 1, I think we will still encourage expert users to use the individual packages.

Unrelated question: I can across [this blog post](https://nolanlawson.com/2018/03/20/smaller-lodash-bundles-with-webpack-and-babel/) recently, and it explains how `import { times } from 'lodash'` is worse than `import times from 'lodash/times'` when bundling Lodash with webpack.

It sounds like we won’t have that issue, so my question is, do we avoid that by exporting the modules like this from `can/es`?

```js
export {
	default as connect
} from 'can-connect/all';
```

**Comment by justinbmeyer** [comment]
@matthewp why does #1 do that?  Are you talking about tree-shaking within the individual modules?  This is mostly just a problem for can-connect I think right?  Nothing else really "needs" tree-shaking within a module.

I think we should start with #2.  It is probably easier to accomplish, makes adding new packages much easier, and over-all less to worry about.  Once `ES` has "taken over".  Speaking of which, where is node `ES` support?

**Comment by matthewp** [comment]
@justinbmeyer No, I'm not talking about adding granular exports for packages. I'm just talking about enabling people to get the benefits of tree shaking even if they are not using `can`. For example, this module:

```js
import Component from 'can-component';
import connect from 'can-connect';

console.log(Component);
```

Doesn't use `connect`, but it will still be included because it is a CommonJS module. However if the code was:

```js
import Component from 'can-component/es';
import connect from 'can-connect/es';

console.log(Component);
```

Then `connect` would be removed. It's the same deal as [this](https://github.com/canjs/canjs/issues/4013#issuecomment-376655894) where the wrapper is needed to get tree-shaking.

**Comment by justinbmeyer** [comment]
@matthewp gotcha.  I'd still say most linters would pick up on that example. I still don't think it's worth the extra hassle.  I'd prefer to wait to update everything to `ES` than keep supporting two module syntaxes everywhere.
